wrappers/services.go

```diff
@@ -375,120 +376,202 @@ func userDaemonReload() error {
+		for file, state := range modifiedUnits {
+			if state == nil {
+				// we don't have anything to rollback to, so just remove the
+				// file
+				if e := os.Remove(file); e != nil {
+					inter.Notify(fmt.Sprintf("while trying to remove %s due to previous failure: %v", file, e))
+				}
+			} else {
+				// rollback the file to the previous state
+				if e := osutil.EnsureFileState(file, *state); e != nil {
+					inter.Notify(fmt.Sprintf("while trying to rollback %s due to previous failure: %v", file, e))
+				}
 			}
 		}
-		if writtenSystem && !preseeding {
+		if modifiedSystem && !preseeding {
 			if e := sysd.DaemonReload(); e != nil {
 				inter.Notify(fmt.Sprintf("while trying to perform systemd daemon-reload due to previous failure: %v", e))
 			}
 		}
-		if writtenUser && !preseeding {
+		if modifiedUser && !preseeding {
 			if e := userDaemonReload(); e != nil {
 				inter.Notify(fmt.Sprintf("while trying to perform user systemd daemon-reload due to previous failure: %v", e))
 			}
 		}
 	}()
 
-	// create services first; this doesn't trigger systemd
-	for _, app := range s.Apps {
-		if !app.IsService() {
-			continue
+	for s, opts := range snaps {
+		if s.Type() == snap.TypeSnapd {
+			return nil, fmt.Errorf("internal error: adding explicit services for snapd snap is unexpected")
 		}
-		// Generate service file
-		content, err := generateSnapServiceFile(app, opts)
-		if err != nil {
-			return err
-		}
-		svcFilePath := app.ServiceFile()
-		os.MkdirAll(filepath.Dir(svcFilePath), 0755)
-		if err := osutil.AtomicWriteFile(svcFilePath, content, 0644, 0); err != nil {
-			return err
-		}
-		written = append(written, svcFilePath)
-		switch app.DaemonScope {
-		case snap.SystemDaemon:
-			writtenSystem = true
-		case snap.UserDaemon:
-			writtenUser = true
+
+		if opts == nil {
+			opts = &AddSnapServicesOptions{}
 		}
 
-		// Generate systemd .socket files if needed
-		var socketFiles *map[string][]byte
-		socketFiles, err = generateSnapSocketFiles(app)
-		if err != nil {
-			return err
+		// TODO: do we still need this?
+		if opts.Preseeding {
+			preseeding = true
 		}
-		for path, content := range *socketFiles {
-			os.MkdirAll(filepath.Dir(path), 0755)
-			if err = osutil.AtomicWriteFile(path, content, 0644, 0); err != nil {
-				return err
+
+		for _, app := range s.Apps {
+			if !app.IsService() {
+				continue
 			}
-			written = append(written, path)
-		}
 
-		if app.Timer != nil {
+			handleFile := func(path string, desiredContent []byte) error {
+				// get the existing file state (if any) of the service to have
+				// something to rollback to if we have any errors
+
+				// note we can't use FileReference here since we may be modifying
+				// the file, and the FileReference.State() wouldn't be evaluated
+				// until _after_ we attempted modification
+				oldFileState := osutil.MemoryFileState{}
+				haveOldFile := false
+				st, stErr := os.Stat(path)
+				if stErr == nil {
```

Comment:
this should check if it's file not existing error vs not?

Reply:
why? there are three cases for `os.Stat`:

1. the file exists and we can `stat` it, in which case we try to read it and if perms are wrong then we get an error there
2. the file doesn't exist in which case we get `err != nil` and we have nothing to roll back to
3. there is some other err from stat, which would prevent us from writing the file down below in `osutil.EnsureFileState` and we end up returning the error otherwise

so I don't see the advantage to checking if `stErr` is FileExists etc here since in the case where it's not a file exists we will hit the same error and return just below anyways